connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java

```diff
@@ -458,13 +458,8 @@ boolean sendRecords() {
     }
 
     protected List<SourceRecord> poll() throws InterruptedException {
-        try {
-            return task.poll();
-        } catch (RetriableException | org.apache.kafka.common.errors.RetriableException e) {
-            log.warn("{} failed to poll records from SourceTask. Will retry operation.", this, e);
-            // Do nothing. Let the framework poll whenever it's ready.
-            return null;
-        }
+        retryWithToleranceOperator.reset();
+        return retryWithToleranceOperator.execute(task::poll, Stage.TASK_POLL, this.getClass());
```

Comment:
It looks like the `RetryWithToleranceOperator` currently expects to only be used for per-record kind of operations (conversion, transformation etc.) - [here](https://github.com/apache/kafka/blob/7c3a2846d46f21f2737483eeb7c04e4eee4c2b5f/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java#L217), [here](https://github.com/apache/kafka/blob/7c3a2846d46f21f2737483eeb7c04e4eee4c2b5f/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java#L233) and [here](https://github.com/apache/kafka/blob/7c3a2846d46f21f2737483eeb7c04e4eee4c2b5f/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java#L183) for instance. I think we'll need to make a number of updates there before it can be used in operations such as `SourceTask::poll` and `SinkTask::put`.

Reply:
I think only `total-records-skipped` is a problem which is defined as `Total number of records skipped by this task.`.  `total-record-failures` is actually defined as the total number of failures seen and not the total number of records which have failed. I would assume it should be easy to update the number of records in the poll(), why do you say we need to make a number of updates?

Reply:
`total-record-failures` is defined as the number of "record processing failures" in a task - https://github.com/apache/kafka/blob/7c3a2846d46f21f2737483eeb7c04e4eee4c2b5f/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java#L366-L367. I don't think a failed call to `SourceTask::poll` counts as a record processing failure?

Reply:
The way I understood it is the number of operations that have failed and that's independent of the number of records in the batch. That's what the subtle difference in naming seems to be suggest to me 

```
recordProcessingFailures = createTemplate("total-record-failures", TASK_ERROR_HANDLING_GROUP_NAME,
                "The number of record processing failures in this task.", taskErrorHandlingTags);
recordProcessingErrors = createTemplate("total-record-errors", TASK_ERROR_HANDLING_GROUP_NAME,
                "The number of record processing errors in this task. ", taskErrorHandlingTags);
recordsSkipped = createTemplate("total-records-skipped", TASK_ERROR_HANDLING_GROUP_NAME,
                "The number of records skipped due to errors.", taskErrorHandlingTags);
```
i.e The number of `records` v/s The number of `record processing`. Does that make sense or am I reading too much between the lines?

Reply:
Wondering if [RetryUtil](https://github.com/apache/kafka/blob/0029bc4897e603614a49e0b0f1e623abbe650c61/connect/runtime/src/main/java/org/apache/kafka/connect/util/RetryUtil.java#L55) is a better alternative to use to avoid publishing these metrics. What do you think?